### ISSUE TITLE
Updates on relation field IDs causing issues with subsequent fetching

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#1853]
See internal https://prisma-company.slack.com/archives/C4GCG53BP/p1583514057163400?thread_ts=1583514048.163300&cid=C4GCG53BP

**Comment by dpetrick** [comment]
Different message from the internal discussion with clearer explanation:

There's one big caveat with the current relation-as-IDs implementation: Take the following model and query:
```prisma
// Relation inlining must happen on ModelA
model ModelA {
    b B @id
}
model ModelB {
    id String @default(cuid()) @id
    a A?
}
```
```graphql
mutation {
    updateOneA(
        where: { b: 1 },
        data: {
            b: { connect: { id: 2 } }
        }
    ) {
        b { id }
    }
}
```

The query above does the correct thing but doesn't return the correct data. It returns null (or errors because we got null for a non-null output type), because we're querying a record where the id is assumed to be b: 1 based on the initial fetch during the update. This is not an issue when we update scalar values because we merge them back into the identifier during the update operation, but relations are handled as separate nested queries, which can invalidate the initial ID now as they're part of the ID, in this case the connect writes b: 2 inline into the A record... causing the subsequent result fetch to fail, because it doesn't know that the ID changed from 1 to 2.
The above does NOT apply to relations as part of @@unique and @unique, if, and only if, it's not the primary identifier of the model. 
A primary ID is a query-internal concept only and is not reflected in the data model, but since this issue is about the internal execution model it must be clarified:
For models with an ID definition, the ID will always be the primary ID. For models without ID definitions, the first required field that is @unique is chosen, then @@unique.

This issue can get even more ugly with moer complex queries. They will fail harder and will not execute at all in almost all cases (e.g. they'll roll back):
```prisma
Model A {
    b B
    c C
    d D
    other String
    @@id([b, c, d, other])
}
// ...
```
```graphql
mutation {
    updateOneA(
        where: { b: 1, c: 1, d: 1, other: "test" },
        data: {
            b: { connect: { id: 5 } }
            c: { update: { id: 10 } }
            d: { upsert: {
                where:  { id: 1 }
                create: { id: 15 }
                update: { id: 20 }
            }}
        }
    ) {
        b { id } // -> Expect 5
        c { id } // -> Expect 10
        d { id } // -> Expect either 15 or 20
    }
}
```

Why? Nested queries are run in sequence, the first nested query executes but the second will cause a rollback because the ID changed at best... there are also worst-case scenarios you can construct with this where you're updating a record that is not supposed to be updated with subsequent nested queries.

Tl;dr: Updating inline relations that are part of the primary ID of a model fails in unexpected ways, but does the correct writes under the hood in simple cases. Be extremely careful what you do.

**Comment by pantharshit00** [comment]
ping @janpio Is this still relevant?
